docs/nf-adapt.qmd

```diff
@@ -0,0 +1,107 @@
+# Adaptation with Normalizing Flows
+
+**Experimental and subject to change**
+
+Normalizing flow adaptation through Fisher HMC is a new sampling algorithm that
+automatically reparameterizes a model. It adds some computational cost outside
+model log-density evaluations, but allows sampling from much more difficult
+posterior distributions. For models with expensive log-density evaluations, the
+normalizing flow adaptation can also be much faster, if it can reduce the number
+of log-density evaluations needed to reach a given effective sample size.
+
+The normalizing flow adaptation works by learning a transformation of the parameter
+space that makes the posterior distribution more amenable to sampling. This is done
+by fitting a sequence of invertible transformations (the "flow") that maps the
+original parameter space to a space where the posterior is closer to a standard
+normal distribution. The flow is trained during warmup.
+
+For more information about the algorithm, see the paper todo
+
+Currently, a lot of time is spent on compiling various parts of the normalizing
+flow, and for small models this can take a large amount of the total time.
+Hopefully, we will be able to reduce this overhead in the future.
+
+## Requirements
+
+Install the optional dependencies for normalizing flow adaptation:
+
+```
+pip install 'nutpie[nnflow]'
+```
+
+If you use with PyMC, this will only work if the model is compiled using the jax
+backend, and if the `gradient_backend` is also set to `jax`.
+
+Training of the normalizing flow can often be accelerated by using a GPU (even
+if the model itself is written in Stan, without any GPU support). To enable GPU
+you need to make sure your `jax` installation comes with GPU support, for
+instance by installing it with `pip install 'jax[cuda12]'`, or selecting the
+`jaxlib` version with GPU support, if you are using conda-forge. You can check if
+your installation has GPU support by checking the output of:
+
+```python
+import jax
+jax.devices()
+```
+
+### Usage
+
+To use normalizing flow adaptation in `nutpie`, you need to enable the
+`transform_adapt` option during sampling. Here is an example of how we can use
+it to sample from a difficult posterior:
+
+```{python}
+import pymc as pm
+import nutpie
+import numpy as np
+import arviz
+
+# Define a 100-dimensional funnel model
+with pm.Model() as model:
+    log_sigma = pm.Normal("log_sigma")
+    pm.Normal("x", mu=0, sigma=pm.math.exp(log_sigma / 2), shape=100)
+
+# Compile the model with the jax backend
+compiled = nutpie.compile_pymc_model(
+    model, backend="jax", gradient_backend="jax"
+)
+```
```

Comment:
Perhaps show how to use it via `pm.sample` as well?

Reply:
You can't pass arguments about the normalizing flow through pm.sample right now...

Reply:
I added a small section in the pymc-usage notes about the `pm.sample` api.